src/tree_builder/mod.rs

```diff
@@ -364,6 +364,48 @@ impl<Handle, Sink> TreeBuilder<Handle, Sink>
     pub fn is_fragment(&self) -> bool {
         self.context_elem.is_some()
     }
+
+    fn appropriate_place_for_insertion(&self, override_target: Option<Handle>) -> InsertionPoint<Handle> {
+        use self::tag_sets::*;
+
+        declare_tag_set!(foster_target = table tbody tfoot thead tr);
+        let target = override_target.unwrap_or_else(|| self.current_node());
+        if !(self.foster_parenting && self.elem_in(target.clone(), foster_target)) {
+            // No foster parenting (the common case).
+            return LastChild(target)
+        }
+
+        // Foster parenting
+        // FIXME: <template>
+        let last_table = self.open_elems.iter()
+            .enumerate()
+            .rev()
+            .filter(|&(_, e)| self.html_elem_named(e.clone(), atom!(table)))
+            .next();
+
+        match last_table {
+            None => {
+                LastChild(self.html_elem())
+            }
+            Some((idx, last_table)) => {
+                // Try inserting "inside last table's parent node, immediately before last table"
+                if self.sink.has_parent_node(last_table.clone()) {
+                    BeforeSibling(last_table.clone())
```

Comment:
We've lost the fallback behaviour from https://github.com/servo/html5ever/pull/137/files#diff-3f339094e126866dfc7763b9a1d54beaL232, haven't we?